parquet-hadoop/src/main/java/org/apache/parquet/crypto/keytools/KeyMaterial.java

```diff
@@ -0,0 +1,166 @@
+
+  private final boolean isFooterKey;
+  private final String kmsInstanceID;
+  private final String kmsInstanceURL;
+  private final String masterKeyID;
+  private final boolean isDoubleWrapped;
+  private final String kekID;
+  private final String encodedWrappedKEK;
+  private final String encodedWrappedDEK;
+
+  private KeyMaterial(boolean isFooterKey, String kmsInstanceID, String kmsInstanceURL, String masterKeyID, 
+      boolean isDoubleWrapped, String kekID, String encodedWrappedKEK, String encodedWrappedDEK) {
+    this.isFooterKey = isFooterKey;
+    this.kmsInstanceID = kmsInstanceID;
+    this.kmsInstanceURL = kmsInstanceURL;
+    this.masterKeyID = masterKeyID;
+    this.isDoubleWrapped = isDoubleWrapped;
+    this.kekID = kekID;
+    this.encodedWrappedKEK = encodedWrappedKEK;
+    this.encodedWrappedDEK = encodedWrappedDEK;
+  }
+
+  static KeyMaterial parse(Map<String, String> keyMaterialJson) {
+    boolean isFooterKey = Boolean.valueOf(keyMaterialJson.get(IS_FOOTER_KEY_FIELD));
+    String kmsInstanceID = null;
+    String kmsInstanceURL = null;
+    if (isFooterKey) {
+      kmsInstanceID = keyMaterialJson.get(KMS_INSTANCE_ID_FIELD);
+      kmsInstanceURL = keyMaterialJson.get(KMS_INSTANCE_URL_FIELD);
+    }
+    boolean isDoubleWrapped = Boolean.valueOf(keyMaterialJson.get(DOUBLE_WRAPPING_FIELD));
+    String masterKeyID = keyMaterialJson.get(MASTER_KEY_ID_FIELD);
+    String  encodedWrappedDEK = keyMaterialJson.get(WRAPPED_DEK_FIELD);
+    String kekID = null;
+    String encodedWrappedKEK = null;
+    if (isDoubleWrapped) {
+      kekID = keyMaterialJson.get(KEK_ID_FIELD);
+      encodedWrappedKEK = keyMaterialJson.get(WRAPPED_KEK_FIELD);
+    }
+
+    return new KeyMaterial(isFooterKey, kmsInstanceID, kmsInstanceURL, masterKeyID, isDoubleWrapped, kekID, encodedWrappedKEK, encodedWrappedDEK);
+  }
+
+  static KeyMaterial parse(String keyMaterialString) {
+    Map<String, String> keyMaterialJson = null;
+    try {
+      keyMaterialJson = OBJECT_MAPPER.readValue(new StringReader(keyMaterialString),
+          new TypeReference<Map<String, String>>() {});
+    } catch (IOException e) {
+      throw new ParquetCryptoRuntimeException("Failed to parse key metadata " + keyMaterialString, e);
+    }
+    String keyMaterialType = keyMaterialJson.get(KEY_MATERIAL_TYPE_FIELD);
+    if (!KEY_MATERIAL_TYPE.equals(keyMaterialType)) {
+      throw new ParquetCryptoRuntimeException("Wrong key material type: " + keyMaterialType + 
+          " vs " + KEY_MATERIAL_TYPE);
+    }
+    return parse(keyMaterialJson);
+  }
+
+  static String createSerialized(boolean isFooterKey, String kmsInstanceID, String kmsInstanceURL, String masterKeyID, 
+      boolean isDoubleWrapped, String kekID, String encodedWrappedKEK, String encodedWrappedDEK, boolean isInternalStorage) {
+    Map<String, String> keyMaterialMap = new HashMap<String, String>(10);
+    keyMaterialMap.put(KEY_MATERIAL_TYPE_FIELD, KEY_MATERIAL_TYPE);
+    if (isInternalStorage) {

```

Comment:
This field is serialized in KeyMaterial objects only for internal storage - because in this case, key material and key metadata is the same thing. This field is always deserialized in KeyMetadata objects.

Reply:
Thanks for explaining.